File: wagerproof-ios-native/Wagerproof/Features/Agents/Creation/Step2IdentityView.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels
import WagerproofStores

/// Step 2 of the wizard — name, emoji, color/gradient. Ports
/// `components/agents/creation/Screen2_Identity.tsx`.
///
/// Converted to native Form/Section — the custom `section(title:subtitle:)`
/// helper is gone; all name-validation logic, bindings, and the gradient
/// wire format are preserved verbatim.
struct Step2IdentityView: View {
    @Bindable var store: AgentCreationStore

    /// Gradient color options from RN COLOR_OPTIONS in Screen2_Identity.tsx:29-46.
    /// Stored as "gradient:#hex1,#hex2" — same wire format as Supabase.
    private static let gradients: [String] = [
        "gradient:#6366f1,#ec4899",
        "gradient:#8b5cf6,#06b6d4",
        "gradient:#ef4444,#f97316",
        "gradient:#22c55e,#06b6d4",
        "gradient:#f97316,#eab308",
        "gradient:#ec4899,#8b5cf6",
        "gradient:#06b6d4,#6366f1",
        "gradient:#22c55e,#eab308",
        "gradient:#ef4444,#ec4899",
        "gradient:#8b5cf6,#f97316",
        "gradient:#3b82f6,#22c55e",
        "gradient:#f59e0b,#ef4444",
        "gradient:#14b8a6,#8b5cf6",
        "gradient:#6366f1,#3b82f6",
        "gradient:#dc2626,#7c3aed",
        "gradient:#0ea5e9,#22d3ee",
    ]

    var body: some View {
        Form {
            // Preview row — transparent background so gradient shows through.
            Section {
                VStack(spacing: 12) {
                    ZStack {
                        avatarGradient
                            .frame(width: 80, height: 80)
                            .clipShape(RoundedRectangle(cornerRadius: 24, style: .continuous))
                            .overlay(
                                RoundedRectangle(cornerRadius: 24, style: .continuous)
                                    .strokeBorder(primaryColor, lineWidth: 3)
                            )
                        PixelSpriteAvatar(spriteIndex: AgentSpriteIndex.forSeed(store.draft.name))
                            .padding(6)
                    }

                    Text(store.draft.name.isEmpty ? "Agent Name" : store.draft.name)
                        .font(.system(size: 22, weight: .bold))
                        .foregroundStyle(Color.appTextPrimary)
                        .opacity(store.draft.name.isEmpty ? 0.5 : 1)
                        .lineLimit(1)
                }
                .frame(maxWidth: .infinity)
                .padding(.vertical, 8)
                // Clear row bg so the avatar gradient stands out.
                .listRowBackground(Color.clear)
            }

            // Agent name field — char count + duplicate warning in footer.
            Section {
                TextField("e.g., Sharp Shooter, The Oracle", text: $store.draft.name)
                    .textInputAutocapitalization(.words)
                    .autocorrectionDisabled(true)
                    .font(.system(size: 16, weight: .medium))
            } header: {
                Text("Agent Name")
            } footer: {
                VStack(alignment: .leading, spacing: 4) {
                    HStack {
                        Spacer()
                        Text("\(store.draft.name.count)/50")
                            .font(.system(size: 12))
                            .foregroundStyle(nameOverLimit ? Color(hex: 0xEF4444) : Color.secondary)
                    }
                    if nameIsDuplicate {
                        Text("You already have an agent with this name")
                            .font(.system(size: 13))
                            .foregroundStyle(Color(hex: 0xEF4444))
                    }
                    if store.draft.name.isEmpty {
                        Text("Give your agent a unique name (required)")
                            .font(.system(size: 13))
                            .foregroundStyle(Color.secondary)
                    }
                }
            }

            // Emoji picker — fixed height so paged TabView fits in the row.
            Section {
                SwipeableEmojiPicker(
                    selectedEmoji: $store.draft.avatarEmoji,
                    selectedColor: primaryColor
                )
                .frame(height: 200)
                .listRowInsets(EdgeInsets())
                if store.draft.avatarEmoji.isEmpty {
                    Text("Please select an emoji")
                        .font(.system(size: 13))
                        .foregroundStyle(Color(hex: 0xEF4444))
                }
            } header: {
                Text("Choose an Emoji")
            } footer: {
                Text("Select an emoji to represent your agent (required)")
                    .font(.system(size: 13))
            }

            // Gradient color grid.
            Section {
                colorGrid
                    .listRowInsets(EdgeInsets(top: 12, leading: 12, bottom: 12, trailing: 12))
            } header: {
                Text("Choose a Color")
            } footer: {
                Text("Select a gradient color for your agent")
            }
        }
    }

    // MARK: - Color grid

    private var colorGrid: some View {
        let columns = Array(repeating: GridItem(.flexible(), spacing: 12), count: 4)
        return LazyVGrid(columns: columns, spacing: 12) {
            ForEach(Self.gradients, id: \.self) { gradient in
                let isSelected = store.draft.avatarColor == gradient
                Button {
                    store.draft.avatarColor = gradient
                } label: {
                    ZStack {
                        gradientView(for: gradient)
                            .frame(width: 48, height: 48)
                            .clipShape(Circle())
                        if isSelected {
                            ZStack {
                                Circle()
                                    .fill(Color.white.opacity(0.9))
                                    .frame(width: 24, height: 24)
                                Image(systemName: "checkmark")
                                    .font(.system(size: 12, weight: .bold))
                                    .foregroundStyle(.black)
                            }
                        }
                    }
                    .overlay(
                        Circle()
                            .strokeBorder(isSelected ? Color.white : Color.clear, lineWidth: 3)
                    )
                }
                .buttonStyle(.plain)
            }
        }
    }

    // MARK: - Helpers

    private var primaryColor: Color {
        Color(hexString: primaryHex(from: store.draft.avatarColor)) ?? Color(hex: 0x6366F1)
    }

    private var avatarGradient: some View {
        gradientView(for: store.draft.avatarColor)
    }

    @ViewBuilder
    private func gradientView(for raw: String) -> some View {
        if raw.hasPrefix("gradient:") {
            let stripped = String(raw.dropFirst("gradient:".count))
            let parts = stripped.split(separator: ",")
            let colors: [Color] = parts.compactMap { Color(hexString: String($0)) }
            if colors.count >= 2 {
                LinearGradient(colors: colors, startPoint: .topLeading, endPoint: .bottomTrailing)
            } else if let first = colors.first {
                first
            } else {
                Color(hex: 0x6366F1)
            }
        } else {
            Color(hexString: raw) ?? Color(hex: 0x6366F1)
        }
    }

    private func primaryHex(from raw: String) -> String {
        if raw.hasPrefix("gradient:") {
            let stripped = String(raw.dropFirst("gradient:".count))
            return stripped.split(separator: ",").first.map(String.init) ?? "#6366f1"
        }
        return raw
    }

    private var nameOverLimit: Bool {
        store.draft.name.count > 50
    }

    private var nameIsDuplicate: Bool {
        let trimmed = store.draft.name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !trimmed.isEmpty else { return false }
        return store.existingAgentNames.contains(where: { $0.lowercased() == trimmed })
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/agents/creation/Step2IdentityView.kt
package com.wagerproof.app.features.agents.creation

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextField
import androidx.compose.material3.TextFieldDefaults
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.app.features.agents.AgentColorPalette
import com.wagerproof.app.features.agents.creation.inputs.SwipeableEmojiPicker
import com.wagerproof.core.design.icons.AppIcon
import com.wagerproof.core.design.pixeloffice.PixelSpriteAvatar
import com.wagerproof.core.design.tokens.AppColors
import com.wagerproof.core.models.AgentSpriteIndex
import com.wagerproof.core.stores.AgentCreationStore

private val IdentityGradients = listOf(
    "gradient:#6366f1,#ec4899", "gradient:#8b5cf6,#06b6d4",
    "gradient:#ef4444,#f97316", "gradient:#22c55e,#06b6d4",
    "gradient:#f97316,#eab308", "gradient:#ec4899,#8b5cf6",
    "gradient:#06b6d4,#6366f1", "gradient:#22c55e,#eab308",
    "gradient:#ef4444,#ec4899", "gradient:#8b5cf6,#f97316",
    "gradient:#3b82f6,#22c55e", "gradient:#f59e0b,#ef4444",
    "gradient:#14b8a6,#8b5cf6", "gradient:#6366f1,#3b82f6",
    "gradient:#dc2626,#7c3aed", "gradient:#0ea5e9,#22d3ee",
)

/** Step 2: agent name, pixel character, emoji, and gradient identity. */
@Composable
fun Step2IdentityView(store: AgentCreationStore, modifier: Modifier = Modifier) {
    val draft = store.draft
    val primary = AgentColorPalette.primary(draft.avatarColor)
    val trimmed = draft.name.trim()
    val duplicate = trimmed.isNotEmpty() && store.existingAgentNames.any { it.equals(trimmed, ignoreCase = true) }
    val sprite = draft.spriteIndex ?: AgentSpriteIndex.forSeed(draft.name.ifBlank { "new-agent" })

    Column(
        modifier.fillMaxWidth().background(AppColors.appSurface).verticalScroll(rememberScrollState()).padding(horizontal = 16.dp, vertical = 16.dp),
        verticalArrangement = Arrangement.spacedBy(4.dp),
    ) {
        Column(Modifier.fillMaxWidth().padding(vertical = 10.dp), horizontalAlignment = Alignment.CenterHorizontally) {
            Box(
                Modifier.size(88.dp).clip(RoundedCornerShape(25.dp)).background(Brush.linearGradient(AgentColorPalette.avatarGradient(draft.avatarColor))).border(3.dp, primary, RoundedCornerShape(25.dp)),
                contentAlignment = Alignment.Center,
            ) {
                PixelSpriteAvatar(spriteIndex = sprite, animated = true, modifier = Modifier.size(72.dp))
            }
            Spacer(Modifier.height(10.dp))
            Text(draft.name.ifBlank { "Agent Name" }, color = AppColors.appTextPrimary.copy(alpha = if (draft.name.isBlank()) 0.5f else 1f), fontSize = 22.sp, fontWeight = FontWeight.Bold)
        }

        WizardSectionHeader("AGENT NAME")
        WizardSectionCard {
            TextField(
                value = draft.name,
                onValueChange = { store.draft = store.draft.copy(name = it) },
                modifier = Modifier.fillMaxWidth().semantics { contentDescription = "Agent name" },
                placeholder = { Text("e.g., Sharp Shooter, The Oracle") },
                singleLine = true,
                keyboardOptions = KeyboardOptions(capitalization = KeyboardCapitalization.Words),
                colors = TextFieldDefaults.colors(
                    focusedContainerColor = Color.Transparent,
                    unfocusedContainerColor = Color.Transparent,
                    focusedIndicatorColor = primary,
                    unfocusedIndicatorColor = AppColors.appBorder,
                    focusedTextColor = AppColors.appTextPrimary,
                    unfocusedTextColor = AppColors.appTextPrimary,
                ),
            )
        }
        Row(Modifier.fillMaxWidth().padding(horizontal = 4.dp, vertical = 5.dp)) {
            Text(
                when {
                    duplicate -> "You already have an agent with this name"
                    draft.name.isBlank() -> "Give your agent a unique name (required)"
                    else -> ""
                },
                color = if (duplicate) AppColors.appLoss else AppColors.appTextSecondary,
                fontSize = 12.sp,
                modifier = Modifier.weight(1f),
            )
            Text("${draft.name.length}/50", color = if (draft.name.length > 50) AppColors.appLoss else AppColors.appTextSecondary, fontSize = 12.sp)
        }

        WizardSectionHeader("PIXEL CHARACTER")
        WizardSectionCard {
            Row(Modifier.fillMaxWidth().horizontalScroll(rememberScrollState()), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                repeat(8) { index ->
                    val selected = sprite == index
                    Box(
                        Modifier.size(width = 54.dp, height = 68.dp).clip(RoundedCornerShape(12.dp)).background(if (selected) primary.copy(alpha = 0.18f) else AppColors.appSurfaceMuted).border(if (selected) 2.dp else 1.dp, if (selected) primary else AppColors.appBorder, RoundedCornerShape(12.dp)).clickable { store.draft = store.draft.copy(spriteIndex = index) }.semantics { contentDescription = "Character ${index + 1}${if (selected) ", selected" else ""}" },
                        contentAlignment = Alignment.Center,
                    ) { PixelSpriteAvatar(index, animated = selected, modifier = Modifier.size(width = 44.dp, height = 58.dp)) }
                }
            }
        }

        WizardSectionHeader("CHOOSE AN EMOJI")
        WizardSectionCard {
            SwipeableEmojiPicker(
                selectedEmoji = draft.avatarEmoji,
                onSelect = { store.draft = store.draft.copy(avatarEmoji = it) },
                selectedColor = primary,
                modifier = Modifier.padding(vertical = 8.dp),
            )
        }
        WizardSectionFooter("Select an emoji to represent your agent (required)")

        WizardSectionHeader("CHOOSE A COLOR")
        WizardSectionCard {
            IdentityGradients.chunked(4).forEach { row ->
                Row(Modifier.fillMaxWidth().padding(vertical = 6.dp), horizontalArrangement = Arrangement.SpaceAround) {
                    row.forEach { gradient ->
                        val selected = draft.avatarColor == gradient
                        Box(
                            Modifier.size(50.dp).clip(CircleShape).background(Brush.linearGradient(AgentColorPalette.avatarGradient(gradient))).border(if (selected) 3.dp else 0.dp, Color.White, CircleShape).clickable { store.draft = store.draft.copy(avatarColor = gradient) }.semantics { contentDescription = "Agent color${if (selected) ", selected" else ""}" },
                            contentAlignment = Alignment.Center,
                        ) {
                            if (selected) {
                                Box(Modifier.size(24.dp).clip(CircleShape).background(Color.White), contentAlignment = Alignment.Center) {
                                    Icon(AppIcon.CHECKMARK.imageVector, null, tint = Color.Black, modifier = Modifier.size(13.dp))
                                }
                            }
                        }
                    }
                }
            }
        }
        WizardSectionFooter("Select a gradient color for your agent")
        Spacer(Modifier.height(24.dp))
    }
}
